### ISSUE TITLE
Add comment support in json parsing

### ISSUE THREAD

**Issue by mike-lischke** [svaarala/duktape#1678]
I know comments are not part of the formal specification of JSON code, but it would be so handy to be able to feed e.g. json configuration file content (which contains comments to make clear what entry is for what) to duktape and have this parsed successfully.

So, please add that little extra in the json parser. Thanks.

**Comment by svaarala** [comment]
I could add it to the JX parser without issues and I agree it'd be handy for e.g. configuration files.

Unfortunately the ES `JSON` binding is very explicit in that the syntax must be parsed exactly as is without modifications:

> Conforming implementations of JSON.parse and JSON.stringify must support the exact interchange format described in the ECMA-404 specification without any deletions or extensions to the format.

This is contrary to (most) other parts of the ES spec: you're otherwise almost free to come up with custom syntax, compile any program that ordinarily results in a SyntaxError, etc.

**Comment by svaarala** [comment]
As a clarification: JX is a superset of JSON so JSON+comments will parse as expected. You can also take advantage of some of the niceties like not being required to quote plain ASCII keys:

```
{
    foo: true
}
```

**Comment by mike-lischke** [comment]
Is that JX parser part of duktape?

**Comment by svaarala** [comment]
Yes, see http://duktape.org/guide.html#customjson. Comment syntax is not included at present, however.

**Comment by mike-lischke** [comment]
Ah, `Duktape.enc` is the key here, I see. So, yes, it would be enough if comments where supported in JS and JC.

For usage from duktape/C: is there a direct C call for that or should I use:

```
duk_get_global_string(ctx, "Duktape");
duk_push_string(ctx, "enc");
duk_push_string(ctx, jsondata);
duk_call_prop(ctx, -3, 1);
```
?

**Comment by svaarala** [comment]
There's no C binding at present so you'd need to do something like that.

Adding a C binding wouldn't be difficult as such, but I would like to be able to represent encoding and decoding options in an extensible, non-API-breaking way e.g. as an options object for that API. It's quite easy to foresee that custom JSON variant encoding/decoding will need a set of flags.

Also the Duktape.enc() API is a bit poor in that respect because it mimics the JSON.stringify() and JSON.parse() API for its arguments. I'd rather now use:

```js
Duktape.enc('jx', myValue, {
    replacer: myReplacerFunction,
    indent: 4
});
```

Than:

```js
Duktape.enc('jx', myValue, myReplacerFunction, 4);
```

The former is more readable, and easier to extend without adding more "unnamed" arguments.

**Comment by mike-lischke** [comment]
True, I also prefer using config objects instead of parameters. The code from above worked instantly (after changing "enc" to "dec" for decoding my json config) - very nice. Now I'd only need comment support. When's the next duktape release planned? :-D

**Comment by fatcerberus** [comment]
> Unfortunately the ES JSON binding is very explicit in that the syntax must be parsed exactly as is without modifications

We have Doug Crockford to blame for that.  He was very adamant that JSON should not support comments because they could theoretically be used by a parser as metadata to change the parsing behavior or something like that.  I call BS, but oh well, what can you do.
